File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/products/product_service_unit_controller.py
# erp_prestech_backend/app/Http/Controllers/ProductServiceUnitController.py

import inspect
import logging
from uuid import UUID
from django.shortcuts import render, redirect, get_object_or_404
from django.http import HttpRequest, HttpResponse
from django.core.exceptions import PermissionDenied
from django.urls import reverse
from .._traits.controller import Controller
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import (
    default_permission_denial,
    default_undefined_exception,
)
from ....Models.products.product_service_unit import ProductServiceUnit
from ....Models.products.product_service import ProductService

logger = logging.getLogger(__name__)

class ProductServiceUnitController(Controller):

  @classmethod
  def index(cls, request: HttpRequest) -> HttpResponse:
    REF=f'{cls.__name}::{inspect.currentframe().f_code.co_name}'
    try:
      ctrl = ProductServiceUnitController()
      ctrl.request = request
      ctrl.authorize('manage constant unit')
      units = ProductServiceUnit.objects.filter(
        created_by=request.user.creator_id()
      ).all()
      return render(request,
                    'productServiceUnit/index.html',
                    {'units': units})
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=REF,
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=REF,
        logger=logger
      )

  @classmethod
  def create(cls, request: HttpRequest) -> HttpResponse:
    REF=f'{cls.__name}::{inspect.currentframe().f_code.co_name}'
    try:
      if not request.user.has_perm('create constant unit'):
        raise PermissionDenied('User does not have permission to create constant units.')
      return render(request, 'productServiceUnit/create.html')
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=REF,
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=REF,
        logger=logger
      )
      
  @classmethod
  def store(cls, request: HttpRequest) -> HttpResponse:
    REF=f'{cls.__name}::{inspect.currentframe().f_code.co_name}'
    try:
      ctrl = ProductServiceUnitController() 
      ctrl.request = request
      ctrl.authorize('create constant unit')
      name = request.POST.get('name', '').strip()
      if not name:
        return redirect(get_redirect_url(request))
      if len(name) > 20:
        return redirect(get_redirect_url(request))
      unit = ProductServiceUnit(
        name=name,
        created_by=request.user.creator_id()
      )
      unit.save()
      return redirect(reverse('product-unit.index'))
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=REF,
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=REF,
        logger=logger
      )

  @classmethod
  def edit(cls, request: HttpRequest, id: UUID) -> HttpResponse:
    REF=f'{cls.__name}::{inspect.currentframe().f_code.co_name}'
    try:
      if not request.user.has_perm('edit constant unit'): 
        raise PermissionDenied('User does not have permission to edit constant units')
      unit = get_object_or_404(ProductServiceUnit, pk=id)
      return render(request,
                  'productServiceUnit/edit.html',
                  {'unit': unit})
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=REF,
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=REF,
        logger=logger
      )    

  @classmethod
  def update(cls, request: HttpRequest, id: UUID) -> HttpResponse:
    REF=f'{cls.__name}::{inspect.currentframe().f_code.co_name}'
    try:
      ctrl = ProductServiceUnitController() 
      ctrl.request = request
      ctrl.authorize('edit constant unit')
      unit = get_object_or_404(ProductServiceUnit, pk=id)
      if unit.created_by != request.user.creator_id():
        raise PermissionDenied
      name = request.POST.get('name', '').strip()
      if not name or len(name) > 20:
        return redirect(get_redirect_url(request))
      unit.name = name 
      unit.save()
      return redirect(reverse('product-unit.index'))
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=REF,
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=REF,
        logger=logger
      )

  @classmethod
  def destroy(cls, request: HttpRequest, id: UUID) -> HttpResponse:
    REF=f'{cls.__name}::{inspect.currentframe().f_code.co_name}'
    try:
      ctrl = ProductServiceUnitController() 
      ctrl.request = request
      ctrl.authorize('delete constant unit')
      unit = get_object_or_404(ProductServiceUnit, pk=id)
      if unit.created_by != request.user.creator_id():
        raise PermissionDenied
      if ProductService.objects.filter(unit_id=unit.id).exists():
        return redirect(get_redirect_url(request))
      unit.delete()
      return redirect(reverse('product-unit.index'))
    except PermissionDenied as e:
      return default_permission_denial(
        request, err=e,
        ref=REF,
        logger=logger
      )
    except Exception as e:
      return default_undefined_exception(
        request, err=e,
        ref=REF,
        logger=logger
      )

File: _inc/laravel/app/Http/Controllers/Products/ProductServiceUnitController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    DatabaseConstants,
    PermissionsConstants,
    ViewsConstants
};
use App\Models\{ProductService, ProductServiceUnit};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{
    RedirectResponse,
    Request
};
use Illuminate\Support\Facades\{
    Log,
    Validator
};

class ProductServiceUnitController extends Controller
{
    use ChecksLogin, ChecksPermissions;

    private const PERM_MANAGE = PermissionsConstants::MNG_CT_UNT;
    private const PERM_CREATE = 'create constant unit';
    private const PERM_EDIT  = 'edit constant unit';
    private const PERM_DELETE = 'delete constant unit';
    private const REDIRECT_INDEX = '/';

    public function index(Request $req)
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = ViewsConstants::PRD_SV_UNT . '.' . $fn;

        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        if ($resp = self::guard($req, self::PERM_MANAGE, self::REDIRECT_INDEX)) return $resp;
        try {
            return $this->measureProfile($action, function () use ($userOrRedirect, $view) {
                $units = ProductServiceUnit::where(DatabaseConstants::TABLE_CREATOR, $userOrRedirect->creatorId())->get();
                return view($view, compact('units'));
            });
        } catch (\Throwable $e) {
            Log::error($action . ' failed: ' . $e->getMessage());
            return defaultUndefinedException($req, $e, $action);
        }
    }

    public function create(Request $req)
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = ViewsConstants::PRD_SV_UNT . '.' . $fn;

        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        if ($resp = self::guard($req, self::PERM_CREATE, self::REDIRECT_INDEX)) return $resp;
        return $this->measureProfile($action, fn() => view($view));
    }

    public function store(Request $req): RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        if ($resp = self::guard($req, self::PERM_CREATE, self::REDIRECT_INDEX)) return $resp;
        $v = Validator::make($req->all(), ['name' => 'required|max:20']);
        if ($v->fails()) return redirect()->back()->with('error', $v->errors()->first());
        try {
            return $this->measureProfile($action, function () use ($req, $userOrRedirect) {
                ProductServiceUnit::create([
                    'name'       => $req->input('name'),
                    DatabaseConstants::TABLE_CREATOR => $userOrRedirect->creatorId()
                ]);
                return redirect()->route(ViewsConstants::PRD_SV_UNT . '.index')->with('success', __('Unit successfully created.'));
            });
        } catch (\Throwable $e) {
            Log::error($action . ' failed: ' . $e->getMessage());
            return defaultUndefinedException($req, $e, $action);
        }
    }

    public function show(Request $req, ProductServiceUnit $unit)
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = ViewsConstants::PRD_SV_UNT . '.' . $fn;

        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        if ($resp = self::guard($req, self::PERM_MANAGE, self::REDIRECT_INDEX)) return $resp;
        return $this->measureProfile($action, fn() => view($view, compact('unit')));
    }

    public function edit(Request $req, ProductServiceUnit $unit)
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";
        $view = ViewsConstants::PRD_SV_UNT . '.' . $fn;

        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        if ($resp = self::guard($req, self::PERM_EDIT, self::REDIRECT_INDEX)) return $resp;
        return $this->measureProfile($action, fn() => view($view, compact('unit')));
    }

    public function update(Request $req, ProductServiceUnit $unit): RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        if ($resp = self::guard($req, self::PERM_EDIT, self::REDIRECT_INDEX)) return $resp;
        $v = Validator::make($req->all(), ['name' => 'required|max:20']);
        if ($v->fails()) return redirect()->back()->with('error', $v->errors()->first());
        try {
            return $this->measureProfile($action, function () use ($req, $unit) {
                $unit->update(['name' => $req->input('name')]);
                return redirect()->route(ViewsConstants::PRD_SV_UNT . '.index')->with('success', __('Unit successfully updated.'));
            });
        } catch (\Throwable $e) {
            Log::error($action . ' failed: ' . $e->getMessage());
            return defaultUndefinedException($req, $e, $action);
        }
    }

    public function destroy(Request $req, ProductServiceUnit $unit): RedirectResponse
    {
        $cls = __CLASS__;
        $fn = __FUNCTION__;
        $action = "$cls::$fn";

        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        if ($resp = self::guard($req, self::PERM_DELETE, self::REDIRECT_INDEX)) return $resp;
        if ($unit->created_by !== $userOrRedirect->creatorId())
            return defaultPermissionDenial($req, new AuthorizationException(), $action);
        try {
            return $this->measureProfile($action, function () use ($unit) {
                if (ProductService::where('unit_id', $unit->id)->exists())
                    return redirect()->back()->with('error', __('This unit is already assigned; please reassign or remove related data.'));
                $unit->delete();
                return redirect()->route(ViewsConstants::PRD_SV_UNT . '.index')->with('success', __('Unit successfully deleted.'));
            });
        } catch (\Throwable $e) {
            Log::error($action . ' failed: ' . $e->getMessage());
            return defaultUndefinedException($req, $e, $action);
        }
    }
}
